### PR TITLE
NAS-108291 / 21.04 / Use SPICE instead of VNC for remote display device in VMs

### PR DESCRIPTION
**Pull request by sonicaj**

### CODE REVIEW

**Review by william-gr — COMMENTED** [review]
The idea was to provide both, SPICE and VNC

File: src/middlewared/middlewared/plugins/vm/devices/display.py [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+
+from .device import Device
+from .utils import create_element
+
+
+class DISPLAY(Device):
+
+    schema = Dict(
+        'attributes',
+        Str('resolution', enum=[
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/vm/devices/display.py:17]: Should we have `resolution_choices` API method?

File: src/middlewared/middlewared/plugins/vm/devices/display.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    def web_uri(self, host, password=None):
+        params = [] if self.is_spice_type() else ['autoconnect=1']
+        if self.password_configured():
+            if not password:
+                return
+
+            params.append(f'password={password}')
+
+        get_params = f'?{"&".join(params)}' if params else ''
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/vm/devices/display.py:52]: Should we use `urllib.parse.urlencode`?

File: src/middlewared/middlewared/plugins/vm/devices/display.py [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+        width, height = (attrs.get('resolution') or '1024x768').split('x')
+        return '-s ' + ','.join(filter(
+            bool, [
+                '29',
+                'fbuf',
+                'vncserver',
+                f'tcp={attrs["bind"]}:{attrs["port"]}',
+                f'w={width}',
+                f'h={height}',
+                f'password={attrs["password"]}' if attrs.get('password') else None,
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/vm/devices/display.py:104]: What if `password` contains `,`?

File: src/middlewared/middlewared/plugins/vm/devices/display.py [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+        self.web_process = subprocess.Popen(
+            [
+                '/usr/local/libexec/novnc/utils/websockify/run', '--web', '/usr/local/libexec/novnc/',
+                '--wrap-mode=ignore', start_args['web_bind'], start_args['server_addr']
+            ], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL,
+        )
+
+    def post_stop_vm(self, *args, **kwargs):
+        if self.web_process and psutil.pid_exists(self.web_process.pid):
+            if self.middleware:
```
Comment (themylogin) [inline src/middlewared/middlewared/plugins/vm/devices/display.py:143]: Why can there be two use cases?

**Review by themylogin — CHANGES_REQUESTED** [review]

**Review by themylogin — APPROVED** [review]

**Pull request merged**